testnet/stacks-node/src/tests/signer.rs

```diff
@@ -70,98 +74,545 @@ struct RunningNodes {
+                            panic!("Received Signature ({},{})", &sig.R, &sig.z);
+                        }
+                        OperationResult::SignTaproot(proof) => {
+                            info!("Received SchnorrProof ({},{})", &proof.r, &proof.s);
+                            schnorr_proof = Some(proof);
+                        }
+                        OperationResult::DkgError(dkg_error) => {
+                            panic!("Received DkgError {:?}", dkg_error);
+                        }
+                        OperationResult::SignError(sign_error) => {
+                            panic!("Received SignError {}", sign_error);
+                        }
+                        OperationResult::Dkg(point) => {
+                            panic!("Received aggregate_group_key {point}");
+                        }
+                    }
+                }
+                if schnorr_proof.is_some() || sign_now.elapsed() > timeout {
+                    break;
+                }
+            }
+            let schnorr_proof = schnorr_proof.expect(&format!(
+                "Failed to get schnorr proof signature within {timeout:?}"
+            ));
+            results.push(schnorr_proof);
+        }
+        results
+    }
+
+    fn run_until_epoch_3_boundary(&mut self) {
+        let epochs = self.running_nodes.conf.burnchain.epochs.clone().unwrap();
+        let epoch_3 =
+            &epochs[StacksEpoch::find_epoch_by_id(&epochs, StacksEpochId::Epoch30).unwrap()];
+
+        let epoch_30_boundary = epoch_3.start_height - 1;
+        // advance to epoch 3.0 and trigger a sign round (cannot vote on blocks in pre epoch 3.0)
+        run_until_burnchain_height(
+            &mut self.running_nodes.btc_regtest_controller,
+            &self.running_nodes.blocks_processed,
+            epoch_30_boundary,
+            &self.running_nodes.conf,
+        );
+        info!("Advanced to Nakamoto epoch 3.0 boundary {epoch_30_boundary}! Ready to Sign Blocks!");
+    }
+
+    fn get_current_reward_cycle(&self) -> u64 {
+        let block_height = self
+            .running_nodes
+            .btc_regtest_controller
+            .get_headers_height();
+        self.running_nodes
+            .btc_regtest_controller
+            .get_burnchain()
+            .block_height_to_reward_cycle(block_height)
+            .unwrap()
+    }
+
+    // Will panic if called on a reward cycle that has not had its signers calculated yet
+    fn get_coordinator_sender(&self, reward_cycle: u64) -> &Sender<RunLoopCommand> {
+        debug!(
+            "Getting current coordinator for reward cycle {:?}",
+            reward_cycle
+        );
+        // Calculate which signer is the coordinator
+        let registered_signers_info = &self
+            .stacks_client
+            .get_registered_signers_info(reward_cycle)
+            .unwrap()
+            .unwrap();
+
+        // TODO: do not use the zeroed consensus hash here
```

Comment:
Is there a ticket for this?

Reply:
Yes. https://github.com/stacks-network/stacks-core/issues/4420